drafts/when-elixirs-performance-becomes-rust-y.md

```diff
@@ -0,0 +1,434 @@
+
+We'll also need to add rustler as compiler in the project, and define our Rust crate.
+
+```Elixir
+def project
+  [
+    compilers: [:rustler, :phoenix, :gettext] ++ Mix.compilers(),
+    rustler_crates: [
+      primenumbers: []
+    ],
+    ... # other project options
+  ]
+```
+
+Let's start writing the Rust code! there will be a rust file at `native/primenumbers/src/lib.rs`, which looks like this currently:
+
+```Rust
+use rustler::{Encoder, Env, Error, Term};
+
+mod atoms {
+    rustler_atoms! {
+        atom ok;
+        atom error;
+        //atom __true__ = "true";
+        //atom __false__ = "false";
+    }
+}
+
+rustler::rustler_export_nifs! {
+    "Elixir.PrimeNumbers",
+    [
+        ("add", 2, add)
+    ],
+    None
+}
+
+fn add<'a>(env: Env<'a>, args: &[Term<'a>]) -> Result<Term<'a>, Error> {
+    let num1: i64 = args[0].decode()?;
+    let num2: i64 = args[1].decode()?;
+
+    Ok((atoms::ok(), num1 + num2).encode(env))
+}
+```
+
+The first statement imports the Encoder, Env, Error and Term types from rustler, so we can make use of them.
+The atoms section declares atoms which will map to Elixir's :ok annd :error atoms, so we can provide an interface similar to elixir.
+
+methods defined in the `rustler::rustler_export_nifs!` statement are made available to Elixir, in the PrimeNumbers module. In the example, add is exported and defined with an arity of 2.
+
+The add function decodes the arguments that are provided, adds them together and yields them back to Elixir as `{:ok, <result>}`. (Although I won't dive deeper into it, rustler makes this possible by implementing the [External Term Format](http://erlang.org/doc/apps/erts/erl_ext_dist.html)).
+
+So let's modify that so we can get us some prime numbers.
+
+```Rust
+use rustler::{Encoder, Env, Error, Term};
+
+mod atoms {
+    rustler::rustler_atoms! { // <- We actually prefixed this with rustler! it's a small oversight in the template generation. see https://github.com/rusterlium/rustler/issues/260
+        atom ok;
+        atom error;
+    }
+}
+```
+
+The first part remains the same, we'll declare the ok and error atoms for use with Elixir.
+
+```Rust
+rustler::rustler_export_nifs! {
+    "Elixir.RustNif.PrimeNumbers",
+    [
+        ("prime_numbers", 1, prime_numbers)
+    ],
+    None
+}
+```
+
+We'll define a `prime_numbers` method with an arity of 1 to be exported.
+
+```Rust
+fn is_prime_number(numbers: &Vec<i64>, x: i64) -> bool {
+    if x == 1 {
+        false
+    } else if x == 2 {
+        true
+    } else {
+        !numbers.iter()
+          .any(|&num| x % num == 0)
+    }
+}
+```
+
+We'll define a method to define whether a number is a prime number. It'll be supplied the current list of prime numbers we know, and the number we want to check. We'll define 1 and 2 as separate cases to allow a starting point for us.
```

Comment:
I'm not really following this. Why would you pass the list of currently known prime numbers, if you want to know if a number `x` is a prime number?